### PR TITLE
Improve MortarPeriodicMesh bigtime

### LINKED ISSUE #7045: Improve MortarPeriodicMesh
### Description of the enhancement or error report
Add 3D capabilities and refactor this to utilize libmesh mesh generation methods.

### Rationale for the enhancement or information for reproducing the error
Adds capabilities. 3D and more element types.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

### PR DESCRIPTION
**Pull request by dschwen**
This now inherits from `GeneratedMesh` and supports all options of it (dimensions, element types, biasing, etc.). Much smaller code base as well.

This class augments `GeneratedMesh` with extra _dim-1_ dimensional subdomains that can be used for Mortar interfaces. The mortar interfaces are along the faces perpendicular to user specified cardinal directions. The main application are mortar based periodicity constraints.

Closes #7045

### CODE REVIEW

**Comment by andrsd** [comment]
My plan with assigning the subdomain ID for mortar interfaces was that it will be a user input, i.e. users would say:
```
[Mesh]
  ...
  [./MortarInterface]
    block = <id>
  [../]
```

We have this syntax in place, but we assume it is part of the loaded mesh. However, the goal is to use this information to generate the mortar interfaces. This may or may not fit your needs, but I'd recommend that the user will actually assign the subdomain ID in the input file, i.e. do not use magic numbers to offset the subdomain IDs, let users specify it via an input parameter.

**Comment by andrsd** [comment]
Also, one more thing to be aware of. If you say:
```
[./MortarInterfaces]
  [./left_right]
  [../]
```

The block that is assigned this interface should have subdomain name of `left_right`, so if you need to add an object on that subdomain, you should be able to do so via its human-readable name.

**Comment by dschwen** [comment]
@andrsd my plan is to write an action works with the `MortarPeriodicMesh`. It grabs the mesh and uses the public interface methods of `MortarPeriodicMesh` to get the SubdomainIDs and the periodic directions. The action will then set up the lagrange multiplier and constraints (I'm mostly thinking about gradient periodicity).

I could add a parameter to let the user override the default subdomain ids of 10,11, and 12 though. However this _is_ a class derived from generated mesh, and generated mesh hardcodes the boundary and subdomain IDs as well. I don't see why the user should be allowed to customize these specific ids but not all the other ids.

**Comment by andrsd** [comment]
> I don't see why the user should be allowed to customize these specific ids but not all the other ids.

That suggestion was mostly for the future when we actually get the automated interface generation, then the id listed in the MortarInterafce will be the number assigned to the mesh block...

**Comment by andrsd** [comment]
Actually, when we have the interface generation in place, you will not need this class... Never mind my comments then 😄

File: modules/phase_field/src/mesh/MortarPeriodicMesh.C [hunk 1]
```diff
@@ -89,50 +46,36 @@ MortarPeriodicMesh::clone() const
...
+
+  buildBndElemList();
+
+  // build side meshes
+  for (unsigned short i = 0; i < _dim; ++i)
+    if (_periodic_dirs.contains(i))
+    {
+      BoundaryID current_boundary_id = getBoundaryID(boundary_names[i]);
+
+      for (auto it = bndElemsBegin(); it != bndElemsEnd(); ++it)
```
Comment (friedmud) [inline modules/phase_field/src/mesh/MortarPeriodicMesh.C]: Since you're using C++11 here anyway... this should probably be a range-based-for

Comment (friedmud) [inline modules/phase_field/src/mesh/MortarPeriodicMesh.C]: Oh - I see... we don't expose the range... only the beginning... ending... maybe we should change that.

Comment (dschwen) [inline modules/phase_field/src/mesh/MortarPeriodicMesh.C]: Exactly. That should be the direction those APIs should take now .

**Pull request merged**